[FEAT] 엔티티 설계 및 DB 세팅

PgmJun opened the pull request:
## ✒️ 관련 이슈번호

- Closes #3 

## 🔑 Key Changes
- ERD에 따른 Entity 설계
- QueryDSL 설정
- Redis 세팅
- MySQL 세팅

## 📸 Screenshot


## 📢 To Reviewers
domain/question 패키지에 있는 ElementType은 속성에 대한 이미지URL, 컬러코드를 담고 있는 Enum 객체입니다.
그런데 카드 발급 대기중 이슈로 아직 AWS 계정이 생성되지 않아서 S3를 만들지 못했기 때문에 URL을 입력해둘 수 없었는데요.
이 부분은 3~4일 내로 카드 발급이 완료되면 빠르게 추가입력 해두겠습니다.
우선적으로 코드 내에서도 확인할 수 있도록 `TODO`로 체크해두었습니다.

그리고 제가 코틀린을 첨써봐서 잘못된 부분이나 개선 가능한 사항 있으면 체크 부탁드려욥

PgmJun committed:
✨: feat 엔티티 설계 #3

PgmJun committed:
⚡️: fix 컬럼 네이밍 규칙 수정 #3

PgmJun committed:
✨: feat QueryDSL 설정 #3

PgmJun committed:
✨: feat QueryDSL 추가 설정 #3

PgmJun committed:
✨: feat Redis 설정 #3

PgmJun committed:
✨: feat Redis 설정 #3

PgmJun committed:
⚡️: fix 잘못된 Entity 클래스 import 수정 #3

orijoon98 commented on src/main/kotlin/com/th/plu/domain/member/Onboarding.kt:
```diff
@@ -0,0 +1,27 @@
+    @OneToOne(fetch = FetchType.LAZY, orphanRemoval = true, cascade = [CascadeType.ALL])
+    @JoinColumn(name = "member_id", nullable = false)
+    private var member: Member,
+
+
+
+) {
+}
```
BaseEntity 빠졌어용

PgmJun replied:
> BaseEntity 빠졌어용

헉 감사합니다! 굿굿

orijoon98 commented on src/main/kotlin/com/th/plu/domain/member/Setting.kt:
```diff
@@ -0,0 +1,30 @@
+    @OneToOne(fetch = FetchType.LAZY, orphanRemoval = true, cascade = [CascadeType.ALL])
+    @JoinColumn(name = "member_id", nullable = false)
+    private var member: Member
+
+) : BaseEntity(LocalDateTime.now(), LocalDateTime.now()) {
+}
```
이거 생성자에 now 넣어줘야하나요?

PgmJun replied:
> 이거 생성자에 now 넣어줘야하나요?

그런 것 같아요 안그럼 빨간 줄 뜨더라구요..? 좀 더 찾아볼게용!

PgmJun replied:
<img width="594" alt="스크린샷 2023-12-27 오후 4 28 58" src="https://github.com/Team-Plu/Plu-Server/assets/84304802/2e0b692c-2125-4786-946d-1560bcd7d1e1">

PgmJun replied:
```kotlin
@MappedSuperclass
@EntityListeners(AuditingEntityListener::class)
open class BaseEntity(

    @CreatedDate
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
    @Column(name = "created_at", nullable = false)
    private val createdAt: LocalDateTime? = null,

    @LastModifiedDate
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
    @Column(name = "modified_at", nullable = false)
    private val modifiedAt: LocalDateTime? = null,
) {
}
```

아 이렇게 설정하면 된다는 거 같네요! null로 한번 초기화 해놓으면 되는 가봅니다!

PgmJun replied:
반영완료됐습니다:)

orijoon98 replied:
```
@MappedSuperclass
@EntityListeners(AuditingEntityListener::class)
open class BaseEntity(

    @CreatedDate
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
    @Column(name = "created_at", nullable = false)
    private val createdAt: LocalDateTime = LocalDateTime.now(),

    @LastModifiedDate
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
    @Column(name = "modified_at", nullable = false)
    private val modifiedAt: LocalDateTime = LocalDateTime.now(),
) {

  @PreUpdate
  fun onUpdate() {
    this.modifiedAt = LocalDateTime.now()
  }
}
```

이런식으로 하는거 같아유!

orijoon98 replied:
타입을 nonNullable 로 하는게 어떨까용!

PgmJun replied:
> ```
> @MappedSuperclass
> @EntityListeners(AuditingEntityListener::class)
> open class BaseEntity(
> 
>     @CreatedDate
>     @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
>     @Column(name = "created_at", nullable = false)
>     private val createdAt: LocalDateTime = LocalDateTime.now(),
> 
>     @LastModifiedDate
>     @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
>     @Column(name = "modified_at", nullable = false)
>     private val modifiedAt: LocalDateTime = LocalDateTime.now(),
> ) {
> 
>   @PreUpdate
>   fun onUpdate() {
>     this.modifiedAt = LocalDateTime.now()
>   }
> }
> ```
> 
> 이런식으로 하는거 같아유!

@PreUpdate는 엔티티에 수정이 발생했을 때, 작동하는 메서드로 알고 있는데용
@LastModifiedDate가 있으니까 @PreUpdate는 없어도 괜찮지 않을까요??

PgmJun replied:
> 타입을 nonNullable 로 하는게 어떨까용!

좋습니다!!

daehwan2da commented on build.gradle.kts:
```diff
@@ -26,15 +28,33 @@ repositories {
+
+	// tools
 	implementation("org.springframework.boot:spring-boot-starter-validation")
+	compileOnly("org.projectlombok:lombok")
```
lombok 은 불필요할듯하네요

PgmJun replied:
> lombok 은 불필요할듯하네요

Constructor의 접근제어 관련해서 
```java
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Builder(access = AccessLevel.PRIVATE)
```

위와 같은 보일러 플레이트 코드를 제거하기 위해 lombok 사용을 선택했는데요
혹시 코틀린은 lombok을 뺀다면 생성자 접근 제어는 어떤 방법으로 처리하나요??

sookyungg replied:
저는 자바 개발을 하기로해서 조금 찾아봤는데 ..자바 개발할 때 lombok이 없어도 될까요?

조금 조사해보니 lombok으로 코드를 작성하고 delombok 해줘야하는 과정이 필요할 것 같기도 합니다 

근데 또 찾아보니 밑에 링크보시면 새 버전부터 코틀린에서 lombok 지원해준다능 이야기가 나와있는 것 같아용!

[https://tech.kakaopay.com/post/kotlin-migration/](url)

daehwan2da replied:
kotlin 에서 애초에 no, all argu, getter, setter 등이 디폴트로 제공됩니다
(data class 를 쓰는 경우에는 equalsAndHashCode까지)
lombok 이 편리함을 주긴하는데, 보일러 플레이트코드가 반드시 코드상에서 제거되어야하는지를 한번 고민해보면 좋을것같네요

- java 에서는 lombok 쓰는건 상관없을것같네요
  - 이전에 kotlin 에서는 java 보다 먼저 compile 되면서 lombok을 씹던가.. 이유는 까먹음..ㅎ
  - 추가로) 새버전에서 지원해주는건 좋은것같네요 ㅎㅎ 

아래는 생성자 접근제어 방식

<img width="1041" alt="스크린샷 2023-12-28 오전 10 31 55" src="https://github.com/Team-Plu/Plu-Server/assets/88091704/e5c21d52-1e7e-4034-aa2c-905d739b01c3">

sookyungg replied:
오! 그럼 제가 작성하는 부분은 lombok이 필요할 수 있으니 남겨둬주세욥...

문제 발생시 제가 다 없애겠습니다..ㅎㅎ

PgmJun replied:
> kotlin 에서 애초에 no, all argu, getter, setter 등이 디폴트로 제공됩니다 (data class 를 쓰는 경우에는 equalsAndHashCode까지) lombok 이 편리함을 주긴하는데, 보일러 플레이트코드가 반드시 코드상에서 제거되어야하는지를 한번 고민해보면 좋을것같네요
> 
> * java 에서는 lombok 쓰는건 상관없을것같네요
>   
>   * 이전에 kotlin 에서는 java 보다 먼저 compile 되면서 lombok을 씹던가.. 이유는 까먹음..ㅎ
>   * 추가로) 새버전에서 지원해주는건 좋은것같네요 ㅎㅎ
> 
> 아래는 생성자 접근제어 방식
> 
> <img alt="스크린샷 2023-12-28 오전 10 31 55" width="1041" src="https://private-user-images.githubusercontent.com/88091704/293119613-e5c21d52-1e7e-4034-aa2c-905d739b01c3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM4MjI5MjIsIm5iZiI6MTcwMzgyMjYyMiwicGF0aCI6Ii84ODA5MTcwNC8yOTMxMTk2MTMtZTVjMjFkNTItMWU3ZS00MDM0LWFhMmMtOTA1ZDczOWIwMWMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyMzEyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI5VDA0MDM0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlNjRjYTAyOWQzY2FiOGU4OGExMWJhZmRmNjk1Njc2MzFiZGY1MjBiYTY5YjVhZGIwMjg1NmViMWFmN2QxZTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MGZf2ixVIRx7_S5TXXB8wUrZlnAy8I3gTXUBe-HnHTE">

오웅 저렇게 처리할 수 있군요! 감사합니닷 ㅎㅎ

daehwan2da commented on src/main/kotlin/com/th/plu/api/config/redis/RedisConfig.kt:
redis setting 을 api 넣은 이유는 무엇이슈

PgmJun replied:
> redis setting 을 api 넣은 이유는 무엇이슈

조만간 멀티 모듈로 분리할 생각이기 때문에 refreshtoken을 관리할 redis의 config 클래스를 
auth관련 API 로직을 개발할 api 패키지에 담아놨슈

daehwan2da commented on src/main/kotlin/com/th/plu/domain/answer/Answer.kt:
```diff
@@ -0,0 +1,40 @@
+
+@Table(name = "answers")
+@Entity
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
```
lombok 은 없어도 좋구, kotlin 에서 기본생성자 만드는 방법은 찾아보면 좋을듯하네요 ㅎㅎ

daehwan2da commented on src/main/kotlin/com/th/plu/domain/answer/Answer.kt:
```diff
@@ -0,0 +1,40 @@
+
+    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "answer_id")
+    private var id: Long? = null,
```
```suggestion
    val id: Long? = null
```

pk 값같은 경우는 캡슐화 + reflection 으로 var 보다는 val

PgmJun replied:
> pk 값같은 경우는 캡슐화 + reflection 으로 var 보다는 val

오호 하나 배워갑니다~!

PgmJun replied:
> pk 값같은 경우는 캡슐화 + reflection 으로 var 보다는 val

private도 지워야하나요??

daehwan2da replied:
private 안지우면 외부에서 getter 접근이안됩니다

PgmJun replied:
> private 안지우면 외부에서 getter 접근이안됩니다

kotlin에서 val은 getter 메서드를 만들어준다고 생각했는데, 알아보니 그게 아니라 필드(프로퍼티)에 직접 접근해서 값을 얻어오는 것이었네요! 하나 배워갑니다 감사합니다:)

daehwan2da commented on src/main/kotlin/com/th/plu/domain/answer/Answer.kt:
```diff
@@ -0,0 +1,40 @@
+    @Column(name = "answer_id")
+    private var id: Long? = null,
+
+    @ManyToOne(fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
```
`CascadeType.ALL` 이유는?

- 저는 개인적으로 이런 영속성 전이보다는 수동으로 처리하는 방향이 예상치 못한 쿼리 발생을 막아줄 수 있어서 좋다고 생각합니다

daehwan2da replied:
사실 `@ManyToOne` 같은 JPA 기반 매핑도 다 뺴는거 추천하긴합니다..ㅎ
(직접 수동으로 핸들링하는거 연습해도 좋을듯 ㅎㅎ)

PgmJun replied:
@daehwan2da 말씀하신 두 부분의 특징은 코드 상에 어떤 동작을 수행할 것인지 '정확히' 명시하게 된다는 점인 것 같네요!
프로젝트가 커지면 문제 발생 시 원인을 분석할 때 훨씬 용이한 설계인 것 같네요 적용해보겠습니다:)

그런데 현재 저희 schema.sql에 짜놓은 ddl과 JPA 매핑 방식은 ddl에서 데이터 삭제/수정 등 작업 시에 자율성을 얻기 위해 fk 설정을 제거하여 테이블간에 매핑을 제거하고,
jpa 상에서만 매핑 애노테이션을 통해 매핑해놓은 상태인데요
말씀하신 것 처럼 설정하려면 ddl 상에는 fk 매핑을 해주는 게 좋으려나요??

PgmJun replied:
추가로 cascade, orphanremoval과 같은 설정을 모드 코드에서 빼는 것 보다, 생명주기가 같은 엔티티들 끼리는 걸어줘도 상관없다는 생각이 드는데 어떻게 생각하시나요?
예를 들어 회원 정보를 삭제하면 Onboarding, Setting 등의 데이터도 같이 삭제하게 되니까 이렇게 생명주기가 같은 부분은 괜찮다고 보는데요 어떻게 생각하세요?

daehwan2da replied:
@PgmJun  entity 상에 연관관계를 안맺는다는 의미는 ddl 상에 fk 도 설정을 안하는 방향의 의미였습니다~ 실무에서는 주로 이런식으로 개발을 많이 해서 제안을 드렸을뿐이지, 개발은 편하신대로 해도 좋을것같아요~

daehwan2da commented on src/main/kotlin/com/th/plu/domain/answer/repository/AnswerRepositoryCustom.kt:
```diff
@@ -0,0 +1,7 @@
+import com.th.plu.domain.answer.Answer
+
+interface AnswerRepositoryCustom {
+    fun findAnswerById(id: Long): Answer?
```
id pk 아니에요? 그냥 query dsl 안타고 findById(id:Long) 태우면 안되나용

PgmJun committed:
⚡️: fix 불필요한 보일러 플레이트 제거 #3

PgmJun committed:
♻️: refactor BaseEntity 기본 생성자 사용가능하도록 변경 #3

daehwan2da commented on src/main/kotlin/com/th/plu/domain/member/MemberSocialType.kt:
```diff
@@ -0,0 +1,9 @@
+import lombok.Getter
+
+@Getter
+enum class MemberSocialType(private val value: String) {
```
```suggestion
enum class MemberSocialType(val value: String) {
```

lombok 제거, private 제거하면, 외부에서 불변한 value get 가능합니다

daehwan2da commented on src/main/kotlin/com/th/plu/domain/member/MemberRole.kt:
```diff
@@ -0,0 +1,9 @@
+import lombok.Getter
+
+@Getter
+enum class MemberRole(private val value: String) {
```
```suggestion
enum class MemberRole(val value: String) {
```

lombok 제거, private 제거하면 외부에서 value get 가능합니다

daehwan2da commented on src/main/kotlin/com/th/plu/domain/member/Member.kt:
```diff
@@ -0,0 +1,42 @@
+    private var role: MemberRole,
+
+    @Column(name = "fcm_token", nullable = false, length = 300)
+    private var fcmToken: String,
```
member : fcm 은 1:1 인가요?

- 이기종 유저 (핸드폰 두개에 같은 계정 사용 혹은 휴대폰 교쳬) 은 어떻게 대응하면 좋을까요

daehwan2da commented on src/main/kotlin/com/th/plu/domain/member/repository/MemberRepositoryCustom.kt:
```diff
@@ -0,0 +1,11 @@
+
+interface MemberRepositoryCustom {
+
+    fun findMemberById(id: Long): Member?
```
pk 인데 custom 할 필요가 없을것같아요

daehwan2da commented on src/main/kotlin/com/th/plu/domain/question/ElementType.kt:
```diff
@@ -0,0 +1,13 @@
+package com.th.plu.domain.question
+
+enum class ElementType(
+    private val characterImageUrl: String,
```
full url 로 쓰기보다는, path 만 갖고 domain 은 분리해 관리해보는 방향은 어떠실까요
- ex) `https://naver.com/image.png` 이대로 저장하는것이 아니라 `/image.png` 만 기록
  - domain 은 동적으로 concat 해줄 수 있도록 컴포넌트 추가 구성
  - 이유는, 우리가 s3 를 쓰다가 다른 플랫폼으로 이관하게되었을때 별도의 코드 수정없이 제공되기 위함

필수는 아닙니다

daehwan2da commented on src/main/kotlin/com/th/plu/domain/answer/Answer.kt:
```diff
@@ -0,0 +1,39 @@
+    private var question: Question,
+
+    @Column(name = "answer_content", nullable = false)
+    private var content: String,
```
sql 상 nullable 한데요, 어떤게 맞을까요?

추가로, nullable 하다면 String? 으로 처리부탁드립니다

PgmJun replied:
> sql 상 nullable 한데요, 어떤게 맞을까요?
> 
> 추가로, nullable 하다면 String? 으로 처리부탁드립니다

엇 응답은 NOT_NULL이 맞습니다! SQL 수정하겠습니다 체크 감사해요!

daehwan2da commented on src/main/kotlin/com/th/plu/domain/answer/Answer.kt:
```diff
@@ -0,0 +1,39 @@
+    private var content: String,
+
+    @Column(name = "is_public", nullable = false)
+    private var isPublic: Boolean
```
sql 상 nullable 한데요, 어떤게 맞을까요?

PgmJun replied:
> sql 상 nullable 한데요, 어떤게 맞을까요?

non-nullable 형태가 맞습니다 체크 감사합니다!

daehwan2da commented on src/main/kotlin/com/th/plu/domain/question/Question.kt:
```diff
@@ -0,0 +1,36 @@
+
+    @Column(name = "element_type", nullable = false, length = 30)
+    @Enumerated(EnumType.STRING)
+    private var elementType: ElementType,
```
sql 상 nullable 하네요 확인 부탁드려요

PgmJun replied:
> sql 상 nullable 하네요 확인 부탁드려요

non-nullable 형태가 맞습니다 체크 감사합니다!

daehwan2da commented on src/main/kotlin/com/th/plu/domain/question/Question.kt:
```diff
@@ -0,0 +1,36 @@
+    private var elementType: ElementType,
+
+    @Column(name = "question_title", nullable = false, length = 100)
+    private var title: String,
```
sql 상 nullable 하네요 확인 부탁드려요

PgmJun replied:
> sql 상 nullable 하네요 확인 부탁드려요

non-nullable 형태가 맞습니다 체크 감사합니다!

daehwan2da commented on src/main/kotlin/com/th/plu/domain/question/Question.kt:
```diff
@@ -0,0 +1,36 @@
+    private var title: String,
+
+    @Column(name = "question_content", nullable = false, length = 300)
+    private var content: String,
```
sql 상 nullable 하네요 확인 부탁드려요

PgmJun replied:
> sql 상 nullable 하네요 확인 부탁드려요

non-nullable 형태가 맞습니다 체크 감사합니다!

daehwan2da reviewed (changes requested):
- 전체적으로 클래스에 lombok 이 많이 적용되어있는데요, lombok 의존 자체를 제거해도 좋을것같습니다
- 영속성 전이 (cascade, orphan.. ) 는 의도치 않은 쿼리가 발생할 수 있고, N 건 쿼리 발생도 일어날 수 있다고 생각들어서요 사용은 지양하는게 좋을것같다는게 제 개인 의견입니다 ㅎ 
  - mysql 기반 jpa 를 안쓴지 오래되어서.. 사실 객체관계 (ex. oneToMany) 도 지양하고싶네요

PgmJun committed:
⚡️: fix RepositoryImpl 객체 Bean 등록 #3

PgmJun committed:
⚡️: fix nonNullable 타입으로 변환 #3

PgmJun committed:
⚡️: fix 코드리뷰 반영 #3

- id 컬럼은 불변하도록 val로 선언
- schema의 not null 미설정 문제 수정
- 코틀린 특성을 고려하여 엔티티의 lombok 보일러 플레이트 코드 제거

PgmJun committed:
⚡️: fix getter,setter 사용을 위해 private 접근 제어자 제거 #3

코틀린은 메서드가 아니라 필드 직접 접근 방식으로 데이터에 접근하기 때문에 private으로 설정 시, 외부에서 값을 확인하거나 수정할 수 없게됨. 때문에 private 접근 제어자 제거
